### PR TITLE
fix(electrostatics): avoid per-system backward recomputation

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Add a first-class per-system energy layout to the monopole Ewald, PME, and slab APIs while preserving the existing per-atom default.

Batched training commonly reduces per-atom energies into one total per system. The existing API required callers to perform that reduction themselves; this change adds a direct `energy_reduction="system"` path and improves backward routing for existing manual reductions. JAX exposes the same energy layout at its public wrapper boundaries.

Previously, a batched per-system loss used a manual reduction:

```python
atom_energy = particle_mesh_ewald(
    positions,
    charges,
    cell,
    batch_idx=batch_idx,
)  # shape: (N,)
system_energy = torch.zeros(
    num_systems, dtype=atom_energy.dtype, device=atom_energy.device
)
system_energy.scatter_add_(0, batch_idx, atom_energy)
loss = (system_weights * system_energy).sum()
```

The new API performs the reduction directly:

```python
system_energy = particle_mesh_ewald(
    positions,
    charges,
    cell,
    batch_idx=batch_idx,
    energy_reduction="system",
)  # shape: (B,)
loss = (system_weights * system_energy).sum()
forces = -torch.autograd.grad(loss, positions)[0]
```

A synchronized CUDA benchmark with 10,000 atoms, float64 inputs, and a `128 x 128 x 128` PME mesh measured the old manual pattern at 356.66 ms before the fix. After the fix, the same pattern takes 12.29 ms, while the new `energy_reduction="system"` path takes 11.99 ms. For comparison, directly summing atom energies takes 12.68 ms after the fix.

The manual reduction produces an all-ones per-atom output gradient stored as a regular tensor. The previous CUDA routing could recognize this uniform gradient only when broadcasting was visible from tensor metadata, so the manual pattern unnecessarily recomputed reciprocal space. Eager mode now checks regular uniform gradients exactly, and the explicit system layout carries per-system gradients directly to the cached PME backward without inspecting per-atom values. Both routes matched the reference force result with maximum absolute differences no larger than `2.91e-11`.

The system layout is important beyond eager-mode convenience. During `torch.compile` or CUDA graph capture, inspecting GPU gradient values is unavailable because it would introduce a graph break or a host synchronization inside capture. Consequently, the eager value check cannot prove that a regular per-atom gradient from a manual reduction is uniform, and that path may fall back to reciprocal-space recomputation. With `energy_reduction="system"`, the backward contract is structural: the output gradient already has shape `(B,)`, so arbitrary per-system loss weights go directly to the cached backward without value inspection. This preserves the fast path in compiled and captured workloads.

The default `energy_reduction="atom"` remains compatible with existing `(N,)` energy outputs. Only the energy field changes layout; forces, charge gradients, and virials retain their established shapes.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

None.

## Changes Made

- Add keyword-only `energy_reduction="atom" | "system"` to the Torch and JAX monopole Ewald, PME, and slab entry points.
- Route Torch system-layout gradients through cached derivative paths without inspecting per-atom values, while retaining exact eager handling for materialized uniform atom cotangents.
- Apply differentiable JAX atom-to-system reductions at public wrapper boundaries without changing derivative-output layouts.
- Preserve atom-major Warp buffers and the established registered slab operator schemas.
- Add API-contract, shape, gradient, higher-order, routing, invalid-input, and direct-output regression coverage.
- Update the changelog, migration guidance, API documentation, and PME training example.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Focused Torch/JAX public-API and cotangent-routing validation passed: 49 tests.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/interactions/electrostatics/_util.py [hunk 1]
```diff
@@ -733,41 +836,76 @@ class _InjectChargeGrad(torch.autograd.Function):
...
-        _energy, _charges, charge_grad, batch_idx = inputs
+        _energy, _charges, charge_grad, batch_idx = inputs[:4]
         ctx.save_for_backward(charge_grad)
         ctx.batch_idx = batch_idx
+        ctx.energy_reduction = inputs[4] if len(inputs) > 4 else "atom"
+        if len(inputs) > 5 and inputs[5] is not None:
+            ctx.num_systems = int(inputs[5])
+        elif inputs[5] is None and batch_idx is not None and batch_idx.numel() > 0:
+            # Legacy private 4-argument calls predate the explicit layout contract.
+            ctx.num_systems = int(batch_idx.max().item()) + 1
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/_util.py]: _InjectChargeGrad.apply binds the forward default arguments before invoking setup_context. Therefore legacy four-argument calls still provide seven entries in inputs, with inputs[5] set to None. The existing four-argument regression tests confirm this path. No IndexError is possible through the supported apply entry point.

File: docs/userguide/about/migration.md [hunk 2]
```diff
@@ -6,6 +6,54 @@
 
 This guide lists user-visible migrations by release.
 
+## Unreleased
```
Comment (laserkelvin) [inline docs/userguide/about/migration.md:9]: Pre-empt with 0.4.1 hotfix IMO

File: nvalchemiops/jax/interactions/electrostatics/_utils.py [hunk 3]
```diff
@@ -50,6 +50,125 @@ def _prepare_cell(cell: jax.Array) -> tuple[jax.Array, int]:
     return cell, cell.shape[0]
 
 
+_ENERGY_REDUCTIONS = ("atom", "system")
+
+
+def _validate_energy_reduction(energy_reduction: str) -> None:
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/electrostatics/_utils.py]: Literal type

File: nvalchemiops/jax/interactions/electrostatics/_utils.py [hunk 4]
```diff
@@ -50,6 +50,125 @@ def _prepare_cell(cell: jax.Array) -> tuple[jax.Array, int]:
...
+        )
+
+    bidx = batch_idx.astype(jnp.int32)
+    counts = _per_system_atom_counts(batch_idx, system_values.shape[0], num_atoms)
+    return (system_values / jnp.maximum(counts, 1.0))[bidx]
+
+
+def _apply_energy_reduction(
+    result: jax.Array | tuple[jax.Array, ...],
+    energy_reduction: str,
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/electrostatics/_utils.py]: Literal

File: nvalchemiops/jax/interactions/electrostatics/ewald.py [hunk 5]
```diff
@@ -869,6 +873,8 @@ def ewald_real_space(
     compute_forces: bool = False,
     compute_charge_gradients: bool = False,
     compute_virial: bool = False,
+    *,
+    energy_reduction: str = "atom",
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/electrostatics/ewald.py]: Literal

File: nvalchemiops/jax/interactions/electrostatics/slab.py [hunk 6]
```diff
@@ -1084,3 +1001,81 @@ def compute_slab_correction(
...
+        Slab correction energy: per-atom when ``energy_reduction="atom"``,
+        per-system when ``energy_reduction="system"``.
+    forces : jax.Array, shape (N, 3), optional
+        Per-atom slab force.
+    charge_gradients : jax.Array, shape (N,), optional
+        Per-atom slab charge gradient.
+    virial : jax.Array, shape (B, 3, 3), optional
+        Per-system slab virial tensor.
+    """
+    _validate_energy_reduction(energy_reduction)
```
Comment (laserkelvin) [inline nvalchemiops/jax/interactions/electrostatics/slab.py:1072]: Having to do this wrapping did get me thinking - would it be possible to do the reduction validation and then application as a decorator instead?

Code would be cleaner, but I'm not sure if it'll have `jit`/tracing consequences

File: nvalchemiops/torch/interactions/electrostatics/_util.py
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/_util.py:429]: This is a blocking call - any way we could make it not blocking?

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 7]
```diff
@@ -773,6 +883,43 @@ def ewald_real_space(
...
+            neighbor_shifts,
+            neighbor_matrix,
+            neighbor_matrix_shifts,
+            mask_value,
+            p.shape[0],
+            p.device,
+        )
+        cell_3d = c if c.dim() == 3 else c.unsqueeze(0)
+        system_energies = []
+        for system_idx in range(num_systems):
```
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/ewald.py]: This doesn't seem efficient - we'll end up with a lot of kernel launches no?

File: test/interactions/electrostatics/bindings/torch/test_ewald.py [hunk 8]
```diff
@@ -104,6 +104,347 @@
...
+    )
+
+    assert call_count == 1
+    torch.testing.assert_close(position_grad, ref_position_grad, rtol=1e-5, atol=1e-7)
+    torch.testing.assert_close(charge_grad, ref_charge_grad, rtol=1e-5, atol=1e-7)
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda"])
+@pytest.mark.parametrize("part", ["real", "reciprocal", "full"])
+def test_system_energy_matches_atom_reduction_and_weighted_gradient(device, part):
```
Comment (laserkelvin) [inline test/interactions/electrostatics/bindings/torch/test_ewald.py:180]: Do you need to add a case for `torch.compile`?

File: nvalchemiops/torch/interactions/electrostatics/ewald.py
Comment (laserkelvin) [inline nvalchemiops/torch/interactions/electrostatics/ewald.py]: This should also be updated - please check across the board to make sure all the docs are updated with this change

**Review by laserkelvin — CHANGES_REQUESTED** [review]

**Comment by laserkelvin** [comment]
/ok to test [3346a24](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/132/commits/3346a24ebe73c458617cad0db396945d317b9ac6)

File: nvalchemiops/jax/interactions/electrostatics/_utils.py (hunk 3 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/electrostatics/_utils.py]: Fixed in 1a36c94b08b23781c839bd01c893a4d7bc84d82e.

File: nvalchemiops/jax/interactions/electrostatics/_utils.py (hunk 4 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/electrostatics/_utils.py]: Fixed in 1a36c94b08b23781c839bd01c893a4d7bc84d82e.

File: nvalchemiops/jax/interactions/electrostatics/ewald.py (hunk 5 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/electrostatics/ewald.py]: Fixed in 1a36c94b08b23781c839bd01c893a4d7bc84d82e.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Fixed in 4415105cdeea3e851031a0e524b143e4695be2c5.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py (hunk 7 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Fixed in 2eb7cbb4e821d73eb7125e63f7ad6065d94ea36b. 
This ugliness is resolved, properly moved to Warp kernels.

Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: JAX already performs one differentiable, JIT-compiled scatter after producing atom energies. Leaving as is for now.

File: nvalchemiops/torch/interactions/electrostatics/_util.py
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/_util.py:429]: We cannot. 
We only reach it when tensor metadata can’t prove that the atom cotangent is uniform; `torch.compile`, CUDA graph capture, and the explicit energy_reduction="system" path all bypass it.

File: nvalchemiops/jax/interactions/electrostatics/slab.py (hunk 6 above)
Comment (zubatyuk) [reply nvalchemiops/jax/interactions/electrostatics/slab.py:1072]: A decorator should be fine under jit as long as `energy_reduction` remains static and we preserve the wrapped signature. Other wrappers have multiple return paths and don’t all pass the same cell object, so a shared decorator would add a fair amount of argument and result plumbing.

File: test/interactions/electrostatics/bindings/torch/test_ewald.py (hunk 8 above)
Comment (zubatyuk) [reply test/interactions/electrostatics/bindings/torch/test_ewald.py:180]: Fixed in c5a271b7d9a19662e83af5bef3c3d828a0586c98.

File: docs/userguide/about/migration.md (hunk 2 above)
Comment (zubatyuk) [reply docs/userguide/about/migration.md:9]: Will do follow-up in rc branch.

**Comment by laserkelvin** [comment]
/ok to test [c5a271b](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/132/commits/c5a271b7d9a19662e83af5bef3c3d828a0586c98)

**Review by laserkelvin — APPROVED** [review]

**Comment by laserkelvin** [comment]
All that changed was CHANGELOG that I resolved, don't need to rerun tests
